wom/src/main/scala/wom/util/YamlUtils.scala

```diff
@@ -0,0 +1,102 @@
+package wom.util
+
+import java.io.StringReader
+
+import com.typesafe.config.ConfigException.BadValue
+import com.typesafe.config.{Config, ConfigFactory}
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
+import io.circe.{Json, ParsingFailure}
+import net.ceedubs.ficus.Ficus._
+import net.ceedubs.ficus.readers.ValueReader
+
+import scala.collection.JavaConverters._
+
+object YamlUtils {
+
+  private[util] implicit val refinedNonNegativeReader: ValueReader[Int Refined NonNegative] = {
+    (config: Config, path: String) => {
+      val int = config.getInt(path)
+      refineV[NonNegative](int) match {
+        case Left(error) => throw new BadValue(path, error)
+        case Right(refinedInt) => refinedInt
+      }
+    }
+  }
+
+  private val defaultMaxNodes = ConfigFactory.load().as[Int Refined NonNegative]("yaml.max-nodes")
+
+  /**
+    * Parses the yaml, detecting loops, and a maximum number of nodes.
+    *
+    * See: https://en.wikipedia.org/wiki/Billion_laughs_attack
+    *
+    * @param yaml     The yaml string.
+    * @param maxNodes Maximum number of yaml nodes to parse.
+    * @return The parsed yaml.
+    */
+  def parse(yaml: String, maxNodes: Int Refined NonNegative = defaultMaxNodes): Either[ParsingFailure, Json] = {
+    try {
+      val snakeYamlParser = new org.yaml.snakeyaml.Yaml()
+      val parsed = snakeYamlParser.load[AnyRef](new StringReader(yaml))
+      if (parsed == null) {
+        // Behave just like circe-yaml 0.9.0
+        // https://github.com/circe/circe-yaml/blob/v0.9.0/src/main/scala/io/circe/yaml/parser/package.scala#L79-L98
+        throw new MatchError(parsed)
+      }
+      searchForLoops(parsed, Nil, maxNodes, Array[Int](0))
+      io.circe.yaml.parser.parse(yaml)
+    } catch {
+      case matchError: MatchError =>
+        throw matchError
```

Comment:
Why catch this just to re-throw it (and possibly with a re-set stack trace)?

Looking at the other case, shouldn't this be a `Left(...)` of some kind?